File: ndfs-py/tests/test_endian.py
"""
Tests for ndfs.endian -- big-endian read/write helpers.

SPDX-License-Identifier: MIT
Copyright (c) 1985-2026 Ronny Hansen
HackerCorp Labs - https://github.com/HackerCorpLabs
"""
import pytest
from ndfs.endian import read_uint16_be, read_uint32_be, write_uint16_be, write_uint32_be


# ── read_uint16_be ──────────────────────────────────────────────────


class TestReadUint16BE:
    def test_reads_0x0000(self):
        assert read_uint16_be(bytes([0x00, 0x00]), 0) == 0

    def test_reads_0xFFFF(self):
        assert read_uint16_be(bytes([0xFF, 0xFF]), 0) == 0xFFFF

    def test_reads_0x1234(self):
        assert read_uint16_be(bytes([0x12, 0x34]), 0) == 0x1234

    def test_reads_at_offset(self):
        assert read_uint16_be(bytes([0x00, 0x00, 0xAB, 0xCD]), 2) == 0xABCD


# ── read_uint32_be ──────────────────────────────────────────────────


class TestReadUint32BE:
    def test_reads_0x00000000(self):
        assert read_uint32_be(bytes([0, 0, 0, 0]), 0) == 0

    def test_reads_0xFFFFFFFF(self):
        assert read_uint32_be(bytes([0xFF, 0xFF, 0xFF, 0xFF]), 0) == 0xFFFFFFFF

    def test_reads_0x12345678(self):
        assert read_uint32_be(bytes([0x12, 0x34, 0x56, 0x78]), 0) == 0x12345678

    def test_reads_at_offset(self):
        data = bytes([0x00, 0x00, 0xDE, 0xAD, 0xBE, 0xEF])
        assert read_uint32_be(data, 2) == 0xDEADBEEF


# ── write_uint16_be ─────────────────────────────────────────────────


class TestWriteUint16BE:
    def test_writes_0x1234(self):
        buf = bytearray(2)
        write_uint16_be(buf, 0, 0x1234)
        assert buf[0] == 0x12
        assert buf[1] == 0x34

    def test_round_trips_with_read(self):
        buf = bytearray(4)
        write_uint16_be(buf, 1, 0xABCD)
        assert read_uint16_be(buf, 1) == 0xABCD


# ── write_uint32_be ─────────────────────────────────────────────────


class TestWriteUint32BE:
    def test_writes_0xDEADBEEF(self):
        buf = bytearray(4)
        write_uint32_be(buf, 0, 0xDEADBEEF)
        assert buf[0] == 0xDE
        assert buf[1] == 0xAD
        assert buf[2] == 0xBE
        assert buf[3] == 0xEF

    def test_round_trips_with_read(self):
        buf = bytearray(8)
        write_uint32_be(buf, 2, 0x12345678)
        assert read_uint32_be(buf, 2) == 0x12345678

    def test_handles_high_bit_correctly_unsigned(self):
        buf = bytearray(4)
        write_uint32_be(buf, 0, 0x80000000)
        assert read_uint32_be(buf, 0) == 0x80000000

File: ndfs-c/tests/test_endian.c
/**
 * Tests for endian helpers.
 *
 * SPDX-License-Identifier: MIT
 * Copyright (c) 1985-2026 Ronny Hansen
 */

#include "test_framework.h"
#include "endian_util.h"

static int test_read_u16be(void)
{
    uint8_t data[] = { 0x12, 0x34 };
    TEST_ASSERT_EQUAL_UINT(0x1234, ndfs_read_u16be(data, 0));
    return 0;
}

static int test_read_u32be(void)
{
    uint8_t data[] = { 0xDE, 0xAD, 0xBE, 0xEF };
    TEST_ASSERT_EQUAL_UINT(0xDEADBEEFu, ndfs_read_u32be(data, 0));
    return 0;
}

static int test_write_u16be(void)
{
    uint8_t data[2] = {0};
    ndfs_write_u16be(data, 0, 0xABCD);
    TEST_ASSERT_EQUAL(0xAB, data[0]);
    TEST_ASSERT_EQUAL(0xCD, data[1]);
    return 0;
}

static int test_write_u32be(void)
{
    uint8_t data[4] = {0};
    ndfs_write_u32be(data, 0, 0x01020304u);
    TEST_ASSERT_EQUAL(0x01, data[0]);
    TEST_ASSERT_EQUAL(0x02, data[1]);
    TEST_ASSERT_EQUAL(0x03, data[2]);
    TEST_ASSERT_EQUAL(0x04, data[3]);
    return 0;
}

static int test_roundtrip_u16(void)
{
    uint8_t data[4] = {0};
    ndfs_write_u16be(data, 1, 0x5678);
    TEST_ASSERT_EQUAL_UINT(0x5678, ndfs_read_u16be(data, 1));
    return 0;
}

static int test_roundtrip_u32(void)
{
    uint8_t data[8] = {0};
    ndfs_write_u32be(data, 2, 0xCAFEBABEu);
    TEST_ASSERT_EQUAL_UINT(0xCAFEBABEu, ndfs_read_u32be(data, 2));
    return 0;
}

void run_endian_tests(void)
{
    TEST_SUITE_BEGIN("Endian Tests");
    RUN_TEST(test_read_u16be);
    RUN_TEST(test_read_u32be);
    RUN_TEST(test_write_u16be);
    RUN_TEST(test_write_u32be);
    RUN_TEST(test_roundtrip_u16);
    RUN_TEST(test_roundtrip_u32);
}
